ouroboros-network/demo/chain-sync.hs

```diff
@@ -0,0 +1,816 @@
+genBlockHeader g prevHeader body =
+    header
+  where
+    header = BlockHeader {
+      headerHash     = hashHeader header,
+      headerPrevHash = maybe GenesisHash (BlockHash . headerHash) prevHeader,
+      headerSlot     = maybe 1 (addSlotGap slotGap  . headerSlot) prevHeader,
+      headerBlockNo  = maybe 1 (succ             . headerBlockNo) prevHeader,
+      headerSigner   = expectedBFTSigner (headerSlot header),
+      headerBodyHash = hashBody body
+    }
+    (slotGap, _) = randomR (1,3) g
+
+    expectedBFTSigner :: SlotNo -> BlockSigner
+    expectedBFTSigner (SlotNo n) = BlockSigner (n `mod` 7)
+
+    addSlotGap :: Int -> SlotNo -> SlotNo
+    addSlotGap m (SlotNo n) = SlotNo (n + fromIntegral m)
+
+genBlockBody :: RandomGen g => g -> BlockBody
+genBlockBody g =
+    BlockBody . take len . drop offset $ bodyData
+  where
+    (offset, g') = randomR (0, bodyDataCycle-1)    g
+    (len   , _ ) = randomR (1, bodyDataCycle*10-1) g'
+
+bodyData :: String
+bodyData = concat
+         . replicate 11
+         $ doloremIpsum
+
+bodyDataCycle :: Int
+bodyDataCycle = length doloremIpsum
+
+doloremIpsum :: String
+doloremIpsum =
+  "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam hendrerit\
+  \ nisi sed sollicitudin pellentesque. Nunc posuere purus rhoncus pulvinar\
+  \ aliquam. Ut aliquet tristique nisl vitae volutpat. Nulla aliquet porttitor\
+  \ venenatis. Donec a dui et dui fringilla consectetur id nec massa. Aliquam\
+  \ erat volutpat. Sed ut dui ut lacus dictum fermentum vel tincidunt neque.\
+  \ Sed sed lacinia lectus. Duis sit amet sodales felis. Duis nunc eros,\
+  \ mattis at dui ac, convallis semper risus. In adipiscing ultrices tellus,\
+  \ in suscipit massa vehicula eu."
+
+--
+-- Mock downloaded block heap
+--
+
+-- | This provides an interface to a collection of dowloaded blocks. This is
+-- enough to implement the 'addFetchedBlock' and 'readFetchedBlocks' methods
+-- in the 'BlockFetchConsensusInterface' and related interfaces.
+--
+-- The interface is enough to use in examples and tests.
+--
+data TestFetchedBlockHeap m block = TestFetchedBlockHeap {
+       getTestFetchedBlocks  :: STM (Set (Point block)),
+       addTestFetchedBlock   :: Point block -> block -> m ()
+     }
+
+-- | Make a 'TestFetchedBlockHeap' using a simple in-memory 'Map', stored in an
+-- 'STM' 'TVar'.
+--
+-- This is suitable for examples and tests.
+--
+mkTestFetchedBlockHeap :: [Point BlockHeader]
+                       -> IO (TestFetchedBlockHeap IO BlockHeader)
+mkTestFetchedBlockHeap points = do
+    v <- atomically (newTVar (Set.fromList points))
+    return TestFetchedBlockHeap {
+      getTestFetchedBlocks = readTVar v,
+      addTestFetchedBlock  = \p _b -> atomically (modifyTVar' v (Set.insert p))
+    }
+
+--
+-- Utils
+--
+
+genesisChainFragment :: AF.AnchoredFragment BlockHeader
+genesisChainFragment = AF.Empty (Point 0 GenesisHash)
+
+shiftAnchoredFragment :: HasHeader block
```

Comment:
There is also `AnchoredFragment.anchorNewest`, but that works with `Chain`s. Maybe using `Chain`s here might make things easier, or we could provide a variant of `anchorNewest` that takes an `AnchoredFragment` as input.

Also: the alignment of the arrows is off.

Reply:
I'll leave `shiftAnchoredFragment` as is for now, I'll only fix the alignment. `anchorNewest` based on `AnchoredFragment` might be useful for testing the volitile db.